subiquity/models/filesystem.py

```diff
@@ -646,3 +69,599 @@ def _dehumanize_size(size):
         raise ValueError("'{}': cannot be negative".format(size_in))
 
     return int(num * mpliers[mplier])
+
+
+def id_factory(name):
+    i = 0
+    def factory():
+        nonlocal i
+        r = "%s-%s"%(name, i)
+        i += 1
+        return r
+    return attr.Factory(factory)
+
+
+def asdict(inst):
+    r = collections.OrderedDict()
+    for field in attr.fields(type(inst)):
+        if field.name.startswith('_'):
+            continue
+        v = getattr(inst, field.name)
+        if v:
+            if hasattr(v, 'id'):
+                v = v.id
+            r[field.name] = v
+    return r
+
+# This code is not going to make much sense unless you have read
+# http://curtin.readthedocs.io/en/latest/topics/storage.html. The
+# Disk, Partition etc classes correspond to entries in curtin's
+# storage config list. They are mostly 'dumb data', all the logic is
+# in the FilesystemModel or FilesystemController classes.
+
+
+@attr.s
+class Disk:
+    id = attr.ib(default=id_factory("disk"))
+    type = attr.ib(default="disk")
+    ptable = attr.ib(default='gpt')
+    serial = attr.ib(default=None)
+    path = attr.ib(default=None)
+    model = attr.ib(default=None)
+    wipe = attr.ib(default=None)
+    preserve = attr.ib(default=False)
+    name = attr.ib(default="")
+    grub_device = attr.ib(default=False)
+
+    _partitions = attr.ib(default=attr.Factory(list), repr=False) # [Partition]
+    _fs = attr.ib(default=None, repr=False) # Filesystem
+    _info = attr.ib(default=None)
+
+    @classmethod
+    def from_info(self, info):
+        d = Disk(info=info)
+        d.serial = info.serial
+        d.path = info.name
+        d.model = info.model
+        return d
+
+    @property
+    def available(self):
+        return self.used < self.size
+
+    @property
+    def next_partnum(self):
+        return len(self._partitions) + 1
+
+    @property
+    def size(self):
+        return self._info.size
+
+    @property
+    def used(self):
+        if self._fs is not None:
+            return self.size
+        r = 0
+        for p in self._partitions:
+            r += p.size
+        return r
+
+    @property
+    def free(self):
+        return self.size - self.used
+
+
+@attr.s
+class Partition:
+    id = attr.ib(default=id_factory("part"))
+    type = attr.ib(default="partition")
+    number = attr.ib(default=0)
+    device = attr.ib(default=None) # Disk
+    size = attr.ib(default=None)
+    wipe = attr.ib(default=None)
+    flag = attr.ib(default=None)
+    preserve = attr.ib(default=False)
+
+    _fs = attr.ib(default=None, repr=False) # Filesystem
```

Comment:
Why is `_fs` non-public (single-leading-underscore)?  I think I noticed a reference to it above, which raises a mild red flag.